WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt

```diff
@@ -0,0 +1,220 @@
+                            requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                        }
+                        .setNegativeButton(R.string.cancel) { _, _ -> }
+                        .create()
+                        .show()
+                },
+                showSpeakerNotSupported = {
+                    AlertDialog.Builder(activity)
+                        .setMessage(R.string.no_speaker_supported)
+                        .setPositiveButton(R.string.ok) { _, _ -> }
+                        .create()
+                        .show()
+                }
+            )
+        }
+
+        val lifecycleOwner = LocalLifecycleOwner.current
+
+        requestPermissionLauncher = rememberLauncherForActivityResult(RequestPermission()) {
+            // We ignore the direct result here, since we're going to check anyway.
+            scope.launch {
+                lifecycleOwner.lifecycle.withStateAtLeast(Lifecycle.State.STARTED) {
+                    stateHolder.onAction(AppAction.PermissionResultReturned)
+                }
+            }
+        }
+
+        // Notify the state holder whenever we become started to reset the state
+        LaunchedEffect(stateHolder) {
+            lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                stateHolder.onAction(AppAction.Started)
+            }
+        }
+
+        SpeakerScreen(
+            appState = stateHolder.appState,
+            isPermissionDenied = stateHolder.isPermissionDenied,
+            recordingProgress = stateHolder.recordingProgress,
+            onMicClicked = {
+                scope.launch {
+                    lifecycleOwner.lifecycle.withStateAtLeast(Lifecycle.State.STARTED) {
+                        stateHolder.onAction(AppAction.MicClicked)
+                    }
+                }
+            },
+            onPlayClicked = {
+                scope.launch {
+                    lifecycleOwner.lifecycle.withStateAtLeast(Lifecycle.State.STARTED) {
+                        stateHolder.onAction(AppAction.PlayClicked)
+                    }
+                }
+            },
+            onMusicClicked = {
+                scope.launch {
+                    lifecycleOwner.lifecycle.withStateAtLeast(Lifecycle.State.STARTED) {
+                        stateHolder.onAction(AppAction.MusicClicked)
+                    }
+                }
+            },
+        )
+    }
+}
+
+/**
+ * A helper method to run the given [block] when the state is at least [state].
+ *
+ * This will wait until the given [state] is achieved, and then will run the block.
+ * If the state drops back down while the [block] is running, the [block] will be cancelled.
+ *
+ * This method resumes successfully when the [block] finishes (either normally, or after being cancelled)
+ */
+private suspend fun Lifecycle.withStateAtLeast(state: Lifecycle.State, block: suspend () -> Unit) {
```

Comment:
Would you be able to use the Lifecycle's own method here? https://developer.android.com/reference/androidx/lifecycle/Lifecycle.State#isAtLeast(androidx.lifecycle.Lifecycle.State)

Reply:
Unfortunately no, since `isAtLeast` is a one-time check.

We could use `isAtLeast` to ensure that the state is started when first launching the block, but we wouldn't get any of the cancellation behavior if the state then drops below started.

Reply:
Added comments above, I don't think these checks are necessary. I need to be convinced :D

Reply:
This method and the ones below (i.e. `raceOf` and `awaitEvents`) seem to add unnecessary complexity to this sample

Reply:
Let me know what you think of the latest changes. I removed these custom `withStateAtLeast` checks, in favor of a more direct `onStopped` callback for the state holder, which will cancel ongoing work.

Reply:
The new implementation seems easier to understand thank you! I added an additional comment to that impl that I think could be even simpler. Thank you